Screenshots are not captured if client.end() is not called right after failure.

## Description
I've noticed that in certain cases screenshots are not captured - it seems that it happens when nightwatch is configured using:

```js
end_session_on_fail: false,
skip_testcases_on_fail: false,
```

and `client.end()` is not called directly after failure (e.g. in `after` hook). 

## Sample test
Here is an example of such issue. In the console log details below you will notice that there is no trace of `saveScreenshot` and `screenshot` commands (eg. ` → Running command: saveScreenshot`). If you comment out test named 'Dummy test' and rerun - screenshot will be captured and relavant commands will be present in log. 

<details><summary>sampleTest.js</summary><p>

```js
module.exports = {
    '@tags': ['screenshot'],
  
    'This test should fail and capture screenshot': function(client) {
      client.url('http://google.com');
      client.waitForElementPresent('#thisWontBePresent')
    },

// comment out dummy test below and rerun the test to see that in such setup screens are captured 
    'Dummy test': function(client) {
      client.pause(100)
    },
        
    after(client) {
      client.end();
    }
  };
  ```
</p></details>

<details><summary>nightwatch.conf.js</summary><p>

```js
const chromedriver = require('chromedriver');

module.exports = {
  src_folders: ['test'],
  silent: false,

  end_session_on_fail: false,
  skip_testcases_on_fail: false,

  globals:{
    waitForConditionTimeout: 1000
  },

  test_settings:{
    default:{
      screenshots: {
        enabled: true,
        path: 'screenshots',
        on_failure: true,
        on_error: true
      },

      webdriver: {
        start_process: true,
        port: 9515,
        server_path: chromedriver.path,
      },
    
      desiredCapabilities : {
        browserName : 'chrome',
        chromeOptions: {
          w3c: false
        }
      },
    }
  }
};

```
</p></details>

**Verbose output:**
<details><summary>console log - with dummy test - screenshots not captured </summary><p>

```txt
PS C:\dev\nightwatch-website-tests> npm run test -- --tag=screenshot

> nightwatch-website-tests@0.0.1 test C:\dev\nightwatch-website-tests
> nightwatch "--tag=screenshot"

 Starting ChromeDriver on port 9515...
 ChromeDriver up and running on port 9515 with pid: 1144 (131ms).

[Issues\screenshots] Test Suite
===============================
- Connecting to localhost on port 9515...
   Request POST  /session  
   { desiredCapabilities:
      { browserName: 'chrome',
        platform: 'ANY',
        chromeOptions: { w3c: false },
| Connecting to localhost on port 9515...

DevTools listening on ws://127.0.0.1:55083/devtools/browser/70c8323a-cdcb-4335-869f-495236fcc727
- Connecting to localhost on port 9515...
   Response 200 POST /session (2733ms)
   { sessionId: '66380b1e2571a9b83f05041fd21bc9bb',
     status: 0,
     value:
      { acceptInsecureCerts: false,
        acceptSslCerts: false,
        applicationCacheEnabled: false,
        browserConnectionEnabled: false,
        browserName: 'chrome',
        chrome:
         { chromedriverVersion:
            '81.0.4044.69 (6813546031a4bc83f717a2ef7cd4ac6ec1199132-refs/branch-heads/4044@{#776})',
           userDataDir:
            'C:\\Users\\llis\\AppData\\Local\\Temp\\scoped_dir1144_423881939' },
        cssSelectorsEnabled: true,
        databaseEnabled: false,
        'goog:chromeOptions': { debuggerAddress: 'localhost:55083' },
        handlesAlerts: true,
        hasTouchScreen: false,
        javascriptEnabled: true,
        locationContextEnabled: true,
        mobileEmulationEnabled: false,
        nativeEvents: true,
        networkConnectionEnabled: false,
        pageLoadStrategy: 'normal',
        platform: 'Windows',
        proxy: {},
        rotatable: false,
        setWindowRect: true,
        strictFileInteractability: false,
        takesHeapSnapshot: true,
        takesScreenshot: true,
        timeouts: { implicit: 0, pageLoad: 300000, script: 30000 },
        unexpectedAlertBehaviour: 'ignore',
        version: '81.0.4044.129',
        webStorageEnabled: true,
i Connected to localhost on port 9515 (2797ms).
  Using: chrome (81.0.4044.129) on Windows platform.

 Received session with ID: 66380b1e2571a9b83f05041fd21bc9bb

 → Running [before]:
 → Completed [before].
Running:  This test should fail and capture screenshot

 → Running [beforeEach]:
 → Completed [beforeEach].

 → Running command: url ('http://google.com')
   Request POST  /session/66380b1e2571a9b83f05041fd21bc9bb/url  
   { url: 'http://google.com' }
   Response 200 POST /session/66380b1e2571a9b83f05041fd21bc9bb/url (958ms)
   { sessionId: '66380b1e2571a9b83f05041fd21bc9bb',
     status: 0,
  → Completed command: url ('http://google.com') (970ms)

 → Running command: waitForElementPresent ('#thisWontBePresent')
   Request POST  /session/66380b1e2571a9b83f05041fd21bc9bb/elements  
   { using: 'css selector', value: '#thisWontBePresent' }
   Response 200 POST /session/66380b1e2571a9b83f05041fd21bc9bb/elements (22ms)
   { sessionId: '66380b1e2571a9b83f05041fd21bc9bb',
     status: 0,
     value: [] }
   Request POST  /session/66380b1e2571a9b83f05041fd21bc9bb/elements  
   { using: 'css selector', value: '#thisWontBePresent' }
   Response 200 POST /session/66380b1e2571a9b83f05041fd21bc9bb/elements (22ms)
   { sessionId: '66380b1e2571a9b83f05041fd21bc9bb',
     status: 0,
     value: [] }
   Request POST  /session/66380b1e2571a9b83f05041fd21bc9bb/elements  
   { using: 'css selector', value: '#thisWontBePresent' }
   Response 200 POST /session/66380b1e2571a9b83f05041fd21bc9bb/elements (23ms)
   { sessionId: '66380b1e2571a9b83f05041fd21bc9bb',
     status: 0,
     value: [] }
× Timed out while waiting for element <#thisWontBePresent> to be present for 1000 milliseconds. - expected "found" but got: "not found" (1100ms)
    at Object.This test should fail and capture screenshot (C:\dev\nightwatch-website-tests\test\issues\screenshots.js:6:14)
    at process._tickCallback (internal/process/next_tick.js:68:7) 

  → Completed command: waitForElementPresent ('#thisWontBePresent') (1111ms)
 → Running [afterEach]:
 → Completed [afterEach].

FAILED: 1 assertions failed (2.098s)
Running:  Dummy test

 → Running [beforeEach]:
 → Completed [beforeEach].

 → Running command: pause (100)
  → Completed command: pause (100) (102ms)
 → Running [afterEach]:
 → Completed [afterEach].
No assertions ran.

 → Running [after]:

 → Running command: end ()

 → Running command: session ('delete', [Function])
   Request DELETE  /session/66380b1e2571a9b83f05041fd21bc9bb  
   Response 200 DELETE /session/66380b1e2571a9b83f05041fd21bc9bb (34ms)
   { sessionId: '66380b1e2571a9b83f05041fd21bc9bb',
     status: 0,
     value: null }
  → Completed command: end () (76ms)
  → Completed command: session ('delete', [Function]) (64ms)
 → Completed [after].
_________________________________________________

TEST FAILURE:  1 assertions failed, 0 passed (5.375s)

 × issues\screenshots
 – This test should fail and capture screenshot (2.098s)
   Timed out while waiting for element <#thisWontBePresent> to be present for 1000 milliseconds. - expected "found" but got: "not found" (1100ms)
       at Object.This test should fail and capture screenshot (C:\dev\nightwatch-website-tests\test\issues\screenshots.js:6:14)
       at process._tickCallback (internal/process/next_tick.js:68:7)

 Wrote report file to: tests_output\issues\CHROME_81.0.4044.129_Windows_screenshots.xml.
 Wrote log file to: C:\dev\nightwatch-website-tests\chromedriver.log.
 ChromeDriver process closed.
npm ERR! code ELIFECYCLE
npm ERR! errno 5
npm ERR! nightwatch-website-tests@0.0.1 test: `nightwatch "--tag=screenshot"`
npm ERR! Exit status 5
npm ERR!
npm ERR! Failed at the nightwatch-website-tests@0.0.1 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\llis\AppData\Roaming\npm-cache\_logs\2020-05-06T21_06_57_248Z-debug.log
```
</p></details>

<details><summary>console log - dummy test commented out - screenshots captured </summary><p>

```txt
PS C:\dev\nightwatch-website-tests> npm run test -- --tag=screenshot

> nightwatch-website-tests@0.0.1 test C:\dev\nightwatch-website-tests
> nightwatch "--tag=screenshot"

 Starting ChromeDriver on port 9515...
 ChromeDriver up and running on port 9515 with pid: 12684 (128ms).

[Issues\screenshots] Test Suite
===============================
- Connecting to localhost on port 9515...
   Request POST  /session  
   { desiredCapabilities:
      { browserName: 'chrome',
        platform: 'ANY',
        chromeOptions: { w3c: false },
| Connecting to localhost on port 9515...

DevTools listening on ws://127.0.0.1:55117/devtools/browser/45c2758a-c0d1-4143-b583-13b277f4efa3
- Connecting to localhost on port 9515...
   Response 200 POST /session (2692ms)
   { sessionId: '9d05dfc50187f0152a8a485724b6678d',
     status: 0,
     value:
      { acceptInsecureCerts: false,
        acceptSslCerts: false,
        applicationCacheEnabled: false,
        browserConnectionEnabled: false,
        browserName: 'chrome',
        chrome:
         { chromedriverVersion:
            '81.0.4044.69 (6813546031a4bc83f717a2ef7cd4ac6ec1199132-refs/branch-heads/4044@{#776})',
           userDataDir:
            'C:\\Users\\llis\\AppData\\Local\\Temp\\scoped_dir12684_1085042448' },
        cssSelectorsEnabled: true,
        databaseEnabled: false,
        'goog:chromeOptions': { debuggerAddress: 'localhost:55117' },
        handlesAlerts: true,
        hasTouchScreen: false,
        javascriptEnabled: true,
        locationContextEnabled: true,
        mobileEmulationEnabled: false,
        nativeEvents: true,
        networkConnectionEnabled: false,
        pageLoadStrategy: 'normal',
        platform: 'Windows',
        proxy: {},
        rotatable: false,
        setWindowRect: true,
        strictFileInteractability: false,
        takesHeapSnapshot: true,
        takesScreenshot: true,
        timeouts: { implicit: 0, pageLoad: 300000, script: 30000 },
        unexpectedAlertBehaviour: 'ignore',
        version: '81.0.4044.129',
        webStorageEnabled: true,
i Connected to localhost on port 9515 (2748ms).
  Using: chrome (81.0.4044.129) on Windows platform.

 Received session with ID: 9d05dfc50187f0152a8a485724b6678d

 → Running [before]:
 → Completed [before].
Running:  This test should fail and capture screenshot

 → Running [beforeEach]:
 → Completed [beforeEach].

 → Running command: url ('http://google.com')
   Request POST  /session/9d05dfc50187f0152a8a485724b6678d/url  
   { url: 'http://google.com' }
   Response 200 POST /session/9d05dfc50187f0152a8a485724b6678d/url (981ms)
   { sessionId: '9d05dfc50187f0152a8a485724b6678d',
     status: 0,
     value: null }
  → Completed command: url ('http://google.com') (994ms)

 → Running command: waitForElementPresent ('#thisWontBePresent')
   { using: 'css selector', value: '#thisWontBePresent' }
   Response 200 POST /session/9d05dfc50187f0152a8a485724b6678d/elements (11ms)
   { sessionId: '9d05dfc50187f0152a8a485724b6678d',
     status: 0,
     value: [] }
   Request POST  /session/9d05dfc50187f0152a8a485724b6678d/elements  
   { using: 'css selector', value: '#thisWontBePresent' }
   Response 200 POST /session/9d05dfc50187f0152a8a485724b6678d/elements (22ms)
   { sessionId: '9d05dfc50187f0152a8a485724b6678d',
     status: 0,
     value: [] }
   Request POST  /session/9d05dfc50187f0152a8a485724b6678d/elements  
   { using: 'css selector', value: '#thisWontBePresent' }
   Response 200 POST /session/9d05dfc50187f0152a8a485724b6678d/elements (25ms)
   { sessionId: '9d05dfc50187f0152a8a485724b6678d',
     status: 0,
     value: [] }
× Timed out while waiting for element <#thisWontBePresent> to be present for 1000 milliseconds. - expected "found" but got: "not found" (1081ms)
    at Object.This test should fail and capture screenshot (C:\dev\nightwatch-website-tests\test\issues\screenshots.js:6:14)
    at process._tickCallback (internal/process/next_tick.js:68:7) 

  → Completed command: waitForElementPresent ('#thisWontBePresent') (1085ms)
 → Running [afterEach]:
 → Completed [afterEach].

FAILED: 1 assertions failed (2.087s)
 → Running [after]:

 → Running command: end ()
 Failures in "This test should fail and capture screenshot". Taking screenshot...

 → Running command: saveScreenshot ('C:\dev\nightwatch-website-tests\screenshots\issues\screenshots\This-test-should-fail-and-capture-screenshot_FAILED_May-06-2020-231317-GMT+0200.png', [Function])

 → Running command: screenshot (false, [Function])
   Request GET  /session/9d05dfc50187f0152a8a485724b6678d/screenshot  
   Response 200 GET /session/9d05dfc50187f0152a8a485724b6678d/screenshot (292ms)
   { sessionId: '9d05dfc50187f0152a8a485724b6678d',
     status: 0,
     value: '',
     suppressBase64Data: true }
  → Completed command: screenshot (false, [Function]) (316ms)
  → Completed command: saveScreenshot ('C:\dev\nightwatch-website-tests\screenshots\issues\screenshots\This-test-should-fail-and-capture-screenshot_FAILED_May-06-2020-231317-GMT+0200.png', [Function]) (323ms)

 → Running command: session ('delete', [Function])
   Request DELETE  /session/9d05dfc50187f0152a8a485724b6678d  
   Response 200 DELETE /session/9d05dfc50187f0152a8a485724b6678d (30ms)
   { sessionId: '9d05dfc50187f0152a8a485724b6678d',
     status: 0,
     value: null }
  → Completed command: end () (390ms)
  → Completed command: session ('delete', [Function]) (42ms)
 → Completed [after].
_________________________________________________

TEST FAILURE:  1 assertions failed, 0 passed (5.494s)

 × issues\screenshots
 – This test should fail and capture screenshot (2.087s)
   Timed out while waiting for element <#thisWontBePresent> to be present for 1000 milliseconds. - expected "found" but got: "not found" (1081ms)
       at Object.This test should fail and capture screenshot (C:\dev\nightwatch-website-tests\test\issues\screenshots.js:6:14)
       at process._tickCallback (internal/process/next_tick.js:68:7)

 Wrote report file to: tests_output\issues\CHROME_81.0.4044.129_Windows_screenshots.xml.
 Wrote log file to: C:\dev\nightwatch-website-tests\chromedriver.log.
 ChromeDriver process closed.
npm ERR! code ELIFECYCLE
npm ERR! errno 5
npm ERR! nightwatch-website-tests@0.0.1 test: `nightwatch "--tag=screenshot"`
npm ERR! Exit status 5
npm ERR!
npm ERR! Failed at the nightwatch-website-tests@0.0.1 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\llis\AppData\Roaming\npm-cache\_logs\2020-05-06T21_13_18_430Z-debug.log
```
</p>

## Your Environment
<!--- Include the relevant details your environment -->

| Executable | Version |
| ---: | :--- |
| `nightwatch --version` | 1.3.4 |
| `npm --version`  | 6.9.0 |
| `node --version` | v10.16.3 |

| Browser driver | Version |
| --- | --- |
| NAME | VERSION |
| chromedriver | 81.0.4044.69 |

| OS | Version |
| --- | --- |
| NAME | VERSION |
| Windows 10 | 1607 |

--- a/lib/api/_loaders/assertion-scheduler.js
+++ b/lib/api/_loaders/assertion-scheduler.js
@@ -13,7 +13,7 @@ class AssertionScheduler {
   }
 
   /**
-   * @param {AssertionLoader} instance
+   * @param {AssertionInstance} instance
    * @param {{rescheduleInterval, abortOnFailure, stackTrace, timeout, reporter}} opts
    */
   constructor(instance, opts = {}) {

--- a/lib/api/client-commands/end.js
+++ b/lib/api/client-commands/end.js
@@ -1,5 +1,4 @@
 const EventEmitter = require('events');
-const {Logger, Screenshots} = require('../../utils');
 
 /**
  * Ends the session. Uses session protocol command.
@@ -20,30 +19,6 @@ class End extends EventEmitter {
     const client = this.client;
 
     if (this.api.sessionId) {
-      if (this.testFailuresExist() && this.shouldTakeScreenshot()) {
-        Logger.info(`Failures in "${this.api.currentTest.name}". Taking screenshot...`);
-
-        const fileNamePath = Screenshots.getFileName({
-          testSuite: this.api.currentTest.module,
-          testCase: this.api.currentTest.name
-        }, client.options.screenshots);
-
-        this.api.saveScreenshot(fileNamePath, (result, err) => {
-          if (!err && this.client.transport.isResultSuccess(result))  {
-            const assertions = this.api.currentTest.results.assertions || [];
-            if (assertions.length > 0) {
-              const currentAssertion = assertions[assertions.length-1];
-              if (currentAssertion) {
-                currentAssertion.screenshots = currentAssertion.screenshots || [];
-                currentAssertion.screenshots.push(fileNamePath);
-              }
-            }
-          } else {
-            Logger.warn('Error saving screenshot...', err || result);
-          }
-        });
-      }
-
       this.api.session('delete', result => {
         client.session.clearSession();
         client.setApiProperty('sessionId', null);
@@ -57,20 +32,6 @@ class End extends EventEmitter {
     }
 
     return this.client.api;
-  }
-
-  testFailuresExist() {
-    let currentTest = this.api.currentTest || {};
-
-    if (currentTest.results) {
-      return currentTest.results.errors > 0 || currentTest.results.failed > 0;
-    }
-
-    return null;
-  }
-
-  shouldTakeScreenshot() {
-    return this.api.options.screenshots.enabled && this.api.options.screenshots.on_failure;
   }
 
   complete(callback, result) {

--- a/lib/api/client-commands/saveScreenshot.js
+++ b/lib/api/client-commands/saveScreenshot.js
@@ -21,14 +21,19 @@ class SaveScreenshot extends ClientCommand {
     const {fileName} = this;
 
     this.api.screenshot(this.api.options.log_screenshot_data, function(result) {
-      Screenshots.writeScreenshotToFile(fileName, result.value, err => {
-        if (err) {
-          Logger.error(`Couldn't save screenshot to "${fileName}":`);
-          Logger.error(err);
-        }
+      if(result.value != null && result.status === 0) {
+        Screenshots.writeScreenshotToFile(fileName, result.value, err => {
+          if (err) {
+            Logger.error(`Couldn't save screenshot to "${fileName}":`);
+            Logger.error(err);
+          }
 
+          callback.call(this, result);
+        });
+      } else {
+        Logger.error(`Couldn't save screenshot to "${fileName}":`);
         callback.call(this, result);
-      });
+      }
     });
   }
 

--- a/lib/api/client-commands/saveScreenshot.js
+++ b/lib/api/client-commands/saveScreenshot.js
@@ -18,10 +18,10 @@ const {Screenshots, Logger} = require('../../utils');
  */
 class SaveScreenshot extends ClientCommand {
   performAction(callback) {
-    const {fileName} = this;
+    const {fileName, client} = this;
 
     this.api.screenshot(this.api.options.log_screenshot_data, function(result) {
-      if (result.value != null && result.status === 0) {
+      if (client.transport.isResultSuccess(result)) {
         Screenshots.writeScreenshotToFile(fileName, result.value, err => {
           if (err) {
             Logger.error(`Couldn't save screenshot to "${fileName}":`);

--- a/lib/api/client-commands/saveScreenshot.js
+++ b/lib/api/client-commands/saveScreenshot.js
@@ -21,7 +21,7 @@ class SaveScreenshot extends ClientCommand {
     const {fileName} = this;
 
     this.api.screenshot(this.api.options.log_screenshot_data, function(result) {
-      if(result.value != null && result.status === 0) {
+      if (result.value != null && result.status === 0) {
         Screenshots.writeScreenshotToFile(fileName, result.value, err => {
           if (err) {
             Logger.error(`Couldn't save screenshot to "${fileName}":`);

--- a/lib/testsuite/index.js
+++ b/lib/testsuite/index.js
@@ -441,9 +441,10 @@ class TestSuite {
   }
 
   takeScreenshot() {
+    const fileNamePath = this.getScreenshotFilenamePath();
     const runnable = new Runnable('screenshot', _ => {
       return new Promise((resolve) => {
-        this.api.saveScreenshot(this.getScreenshotFilenamePath(), (result, err) => {
+        this.api.saveScreenshot(fileNamePath, (result, err) => {
           if (!err && this.client.transport.isResultSuccess(result))  {
             const assertions = this.api.currentTest.results.assertions || [];
             if (assertions.length > 0) {
@@ -460,6 +461,7 @@ class TestSuite {
         });
       });
     });
+
     return runnable.run(this.commandQueue); 
 
   }
@@ -497,7 +499,7 @@ class TestSuite {
         // registering non-assert errors
         this.reporter.registerTestError(err);
 
-        if(this.shouldTakeScreenshotOnError()) {
+        if (this.shouldTakeScreenshotOnError()) {
           await this.takeScreenshot();
         }
 
@@ -521,7 +523,7 @@ class TestSuite {
         });
       }
 
-      if(this.shouldTakeScreenshotOnFailure()) {
+      if (this.shouldTakeScreenshotOnFailure()) {
         await this.takeScreenshot();
       }
       // clearing the queue here to avoid continuing with the rest of the testcase,

--- a/lib/testsuite/index.js
+++ b/lib/testsuite/index.js
@@ -8,7 +8,7 @@ const NightwatchAssertError = require('../assertion').AssertionError;
 const SuiteRetries = require('./retries.js');
 const Nightwatch = require('../index.js');
 const Concurrency = require('../runner/concurrency/concurrency.js');
-const {Logger} = require('../utils');
+const {Logger, Screenshots} = require('../utils');
 
 class TestSuite {
   constructor({modulePath, modules, settings, argv, addtOpts = {}}) {
@@ -421,6 +421,49 @@ class TestSuite {
     return this.runCurrentTest(currentTestName);
   }
 
+  isScreenshotEnabled() {
+    return this.settings.screenshots.enabled;
+  }
+
+  shouldTakeScreenshotOnError() {
+    return this.isScreenshotEnabled() && this.settings.screenshots.on_error;
+  }
+
+  shouldTakeScreenshotOnFailure() {
+    return this.isScreenshotEnabled() && this.settings.screenshots.on_failure;
+  }
+
+  getScreenshotFilenamePath() {
+    return Screenshots.getFileName({
+      testSuite: this.api.currentTest.module,
+      testCase: this.api.currentTest.name
+    }, this.settings.screenshots);
+  }
+
+  takeScreenshot() {
+    const runnable = new Runnable('screenshot', _ => {
+      return new Promise((resolve) => {
+        this.api.saveScreenshot(this.getScreenshotFilenamePath(), (result, err) => {
+          if (!err && this.client.transport.isResultSuccess(result))  {
+            const assertions = this.api.currentTest.results.assertions || [];
+            if (assertions.length > 0) {
+              const currentAssertion = assertions[assertions.length-1];
+              if (currentAssertion) {
+                currentAssertion.screenshots = currentAssertion.screenshots || [];
+                currentAssertion.screenshots.push(fileNamePath);
+              }
+            }
+          } else {
+            Logger.warn('Error saving screenshot...', err || result);
+          }
+          resolve();
+        });
+      });
+    });
+    return runnable.run(this.commandQueue); 
+
+  }
+  
   shouldRetrySuite() {
     return !this.reporter.allTestsPassed && this.suiteRetries.shouldRetrySuite();
   }
@@ -454,6 +497,10 @@ class TestSuite {
         // registering non-assert errors
         this.reporter.registerTestError(err);
 
+        if(this.shouldTakeScreenshotOnError()) {
+          await this.takeScreenshot();
+        }
+
         throw err;
       }
 
@@ -474,6 +521,9 @@ class TestSuite {
         });
       }
 
+      if(this.shouldTakeScreenshotOnFailure()) {
+        await this.takeScreenshot();
+      }
       // clearing the queue here to avoid continuing with the rest of the testcase,
       // unless abortOnFailure is set to false
       if (!isAssertionError(err) || err.abortOnFailure) {

--- a/test/src/api/commands/testEnd.js
+++ b/test/src/api/commands/testEnd.js
@@ -84,57 +84,6 @@ describe('.end()', function() {
     }).catch(err => done(err));
   });
 
-  it('client.end() - with screenshot and custom filename format', function (done) {
-    MockServer.addMock({
-      url: '/wd/hub/session/1352110219202/screenshot',
-      method: 'GET',
-      response: JSON.stringify({
-        status: 0,
-        state: 'success',
-        value: '==content'
-      })
-    }, true);
-
-    Nightwatch.initClient({
-      screenshots: {
-        enabled: true,
-        on_failure: true,
-        path: './screens',
-        filename_format({testSuite, testCase, isError, dateObject}) {
-          return `${testSuite}/${testCase}--failed.png`;
-        }
-      }
-    }).then(client => {
-      client.api.currentTest = {
-        module: 'test_module',
-        name: 'test_name',
-        results: {
-          failed: 1,
-          passed: 0
-        }
-      };
-
-      let screenFileName;
-      const saveScreenshot = client.api.saveScreenshot;
-      client.api.saveScreenshot = function (file, callback) {
-        screenFileName = file;
-      };
-      client.api.end(function callback(result) {
-        assert.strictEqual(result.status, 0);
-      });
-
-      client.start(function(err) {
-        try {
-          assert.ok(screenFileName.endsWith('test_module/test_name--failed.png'));
-          client.api.saveScreenshot = saveScreenshot;
-          done(err);
-        } catch (e) {
-          done(e);
-        }
-      });
-    });
-  });
-
   it('client.end() - failures and screenshots disabled', function (done) {
     MockServer.addMock({
       url: '/wd/hub/session/1352110219202/screenshot',

--- a/test/src/runner/testRunnerScreenshotsOutput.js
+++ b/test/src/runner/testRunnerScreenshotsOutput.js
@@ -1,0 +1,313 @@
+const path = require('path');
+const fs = require('fs');
+const assert = require('assert');
+const common = require('../../common.js');
+const MockServer = require('../../lib/mockserver.js');
+const CommandGlobals = require('../../lib/globals/commands.js');
+const NightwatchClient = common.require('index.js');
+const rimraf = require('rimraf');
+
+describe('testRunnerScreenshotsOutput', function() {
+  
+  before(function(done) {
+    this.server = MockServer.init();
+    this.server.on('listening', () => {
+      if(fileExistsSync('screenshots/')){
+        rimraf.sync('screenshots/');
+      }
+      done();
+    });
+  });
+
+  afterEach(function(done) {
+    if(fileExistsSync('screenshots/')){
+      rimraf.sync('screenshots/');
+    }
+    done();
+  });
+
+  after(function(done) {
+    CommandGlobals.afterEach.call(this, function() {
+      if(fileExistsSync('screenshots/')){
+        rimraf.sync('screenshots/');
+      }
+      done();
+    });
+  });
+
+  beforeEach(function() {
+    process.removeAllListeners('exit');
+    process.removeAllListeners('uncaughtException');
+    process.removeAllListeners('unhandledRejection');
+  });
+
+  it('takes screenshot on each test failure', function () {
+
+
+    let testsPath = [
+      path.join(__dirname, '../../sampletests/withfailures')
+    ];
+
+    let settings = {
+      skip_testcases_on_fail: false,
+      selenium: {
+        port: 10195,
+        version2: true,
+        start_process: true
+      },
+      output_folder: 'output',
+      silent: true,
+      globals: {
+        waitForConditionPollInterval: 5,
+        waitForConditionTimeout: 5,
+        retryAssertionTimeout: 1,
+        reporter: function () {
+        }
+      },
+      output: false,
+      screenshots: {
+        enabled: true,
+        on_failure: true,
+        on_error: true,
+        path: 'screenshots'
+      }
+    };
+
+    MockServer.addMock({
+      url : '/wd/hub/session/1352110219202/screenshot',
+      method:'GET',
+      response : JSON.stringify({
+        sessionId: '1352110219202',
+        status:0,
+        value:'screendata'
+      })
+    }, true);
+
+    return NightwatchClient.runTests(testsPath, settings)
+      .then(_ => {
+        return readDirPromise('screenshots/sample');
+      })
+      .then(files => {
+        assert.ok(files);
+        assert.deepStrictEqual(files.length, 1);
+        files.forEach((val) => assert(val.includes('.png')));
+      });
+  });
+
+  it('takes screenshot for failed test and exits if skip_testcases_on_fail is set to true', function () {
+
+    let testsPath = [
+      path.join(__dirname, '../../sampletests/withfailures')
+    ];
+
+    let settings = {
+      skip_testcases_on_fail: true,
+      selenium: {
+        port: 10195,
+        version2: true,
+        start_process: true
+      },
+      output_folder: 'output',
+      silent: true,
+      globals: {
+        waitForConditionPollInterval: 5,
+        waitForConditionTimeout: 5,
+        retryAssertionTimeout: 1,
+        reporter: function () {
+        }
+      },
+      output: false,
+      screenshots: {
+        enabled: true,
+        on_failure: true,
+        on_error: true,
+        path: 'screenshots'
+      }
+    };
+
+    MockServer.addMock({
+      url : '/wd/hub/session/1352110219202/screenshot',
+      method:'GET',
+      response : JSON.stringify({
+        sessionId: '1352110219202',
+        status:0,
+        value:'screendata'
+      })
+    });
+
+
+    return NightwatchClient.runTests(testsPath, settings)
+      .then(_ => {
+        return readDirPromise('screenshots/sample')
+          .then(files => {
+            assert.ok(files);
+            assert.strictEqual(files.length, 1);
+            files.forEach((val) => assert(val.includes('.png')));
+          });
+      });
+  });
+
+  it('doesnt save file if screenshot call is failed', function () {
+
+    let testsPath = [
+      path.join(__dirname, '../../sampletests/withfailures')
+    ];
+
+    let settings = {
+      skip_testcases_on_fail: true,
+      selenium: {
+        port: 10195,
+        version2: true,
+        start_process: true
+      },
+      output_folder: 'output',
+      silent: true,
+      globals: {
+        waitForConditionPollInterval: 5,
+        waitForConditionTimeout: 5,
+        retryAssertionTimeout: 1,
+        reporter: function () {
+        }
+      },
+      output: false,
+      screenshots: {
+        enabled: true,
+        on_failure: true,
+        on_error: true,
+        path: 'screenshots'
+      }
+    };
+
+    MockServer.addMock({
+      url : '/wd/hub/session/1352110219202/screenshot',
+      method:'GET',
+      response : JSON.stringify({
+        status: -1,
+        state: '',
+        code: '',
+        value: null,
+        errorStatus: '',
+        error: 'Unknown command 404 Not Found',
+        httpStatusCode: 404
+      })
+    });
+
+
+    return NightwatchClient.runTests(testsPath, settings)
+      .then(_ => {
+        return readDirPromise('screenshots/sample')
+          .catch(err => {
+            assert.ok(err);
+            assert.ok(err.message.includes('no such file or directory'));
+          });
+      });
+  });
+
+  it('does not take screenshot if screenshot is disabled', function () {
+
+    let testsPath = [
+      path.join(__dirname, '../../sampletests/withfailures')
+    ];
+
+    let settings = {
+      skip_testcases_on_fail: true,
+      selenium: {
+        port: 10195,
+        version2: true,
+        start_process: true
+      },
+      output_folder: 'output',
+      silent: true,
+      globals: {
+        waitForConditionPollInterval: 5,
+        waitForConditionTimeout: 5,
+        retryAssertionTimeout: 1,
+        reporter: function () {
+        }
+      },
+      output: false,
+      screenshots: {
+        enabled: false,
+        on_failure: true,
+        on_error: true,
+        path: 'screenshots'
+      }
+    };
+
+
+    return NightwatchClient.runTests(testsPath, settings)
+      .then(_ => {
+        return readDirPromise('screenshots/sample')
+          .catch((err) => {
+            assert.ok(err);
+            assert.ok(err.message.includes('no such file or directory'));
+          });
+      });
+  });
+
+  it('does not take screenshot if screenshot is enabled but on_failure is set to false', function () {
+
+    let testsPath = [
+      path.join(__dirname, '../../sampletests/withfailures')
+    ];
+
+    let settings = {
+      skip_testcases_on_fail: true,
+      selenium: {
+        port: 10195,
+        version2: true,
+        start_process: true
+      },
+      output_folder: 'output',
+      silent: true,
+      globals: {
+        waitForConditionPollInterval: 5,
+        waitForConditionTimeout: 5,
+        retryAssertionTimeout: 1,
+        reporter: function () {
+        }
+      },
+      output: false,
+      screenshots: {
+        enabled: true,
+        on_failure: false,
+        on_error: true,
+        path: 'screenshots'
+      }
+    };
+
+
+    return NightwatchClient.runTests(testsPath, settings)
+      .then(_ => {
+        return readDirPromise('screenshots/sample')
+          .catch((err) => {
+            assert.ok(err);
+            assert.ok(err.message.includes('no such file or directory'));
+          });
+      });
+  });
+
+
+});
+
+function readDirPromise(dirName) {
+  return new Promise(function(resolve, reject) {
+    fs.readdir(dirName, function(err, result) {
+      if (err) {
+        return reject(err);
+      }
+
+      resolve(result);
+    });
+  });
+}
+
+// util to replace deprecated fs.existsSync
+function fileExistsSync(path) {
+  try {
+    fs.statSync(path);
+    return true;
+  } catch (e) {
+    return false;
+  }
+}

--- a/test/src/runner/testRunnerScreenshotsOutput.js
+++ b/test/src/runner/testRunnerScreenshotsOutput.js
@@ -14,7 +14,7 @@ describe('testRunnerScreenshotsOutput', function() {
   before(function(done) {
     this.server = MockServer.init();
     this.server.on('listening', () => {
-        rimraf(screenshotFilePath, (error) => {done()});
+      rimraf(screenshotFilePath, (error) => {done()});
     });
   });
 
@@ -24,7 +24,7 @@ describe('testRunnerScreenshotsOutput', function() {
 
   after(function(done) {
     CommandGlobals.afterEach.call(this, function() {
-        rimraf(screenshotFilePath, (error) => {done()});
+      rimraf(screenshotFilePath, (error) => {done()});
     });
   });
 

--- a/test/src/runner/testRunnerScreenshotsOutput.js
+++ b/test/src/runner/testRunnerScreenshotsOutput.js
@@ -14,17 +14,17 @@ describe('testRunnerScreenshotsOutput', function() {
   before(function(done) {
     this.server = MockServer.init();
     this.server.on('listening', () => {
-      rimraf(screenshotFilePath, (error) => {done()});
+      rimraf(screenshotFilePath, done);
     });
   });
 
   afterEach(function(done) {
-    rimraf(screenshotFilePath, (error) => {done()});
+    rimraf(screenshotFilePath, done);
   });
 
   after(function(done) {
     CommandGlobals.afterEach.call(this, function() {
-      rimraf(screenshotFilePath, (error) => {done()});
+      rimraf(screenshotFilePath, done);
     });
   });
 

--- a/test/src/runner/testRunnerScreenshotsOutput.js
+++ b/test/src/runner/testRunnerScreenshotsOutput.js
@@ -5,31 +5,34 @@ const common = require('../../common.js');
 const MockServer = require('../../lib/mockserver.js');
 const CommandGlobals = require('../../lib/globals/commands.js');
 const NightwatchClient = common.require('index.js');
+const utils = common.require('utils/index.js');
 const rimraf = require('rimraf');
 
 describe('testRunnerScreenshotsOutput', function() {
-  
+  let screenshotFilePath = 'screenshots';
+  let moduleName = 'sample';
+
   before(function(done) {
     this.server = MockServer.init();
     this.server.on('listening', () => {
-      if(fileExistsSync('screenshots/')){
-        rimraf.sync('screenshots/');
+      if (utils.dirExistsSync(screenshotFilePath)) {
+        rimraf.sync(screenshotFilePath);
       }
       done();
     });
   });
 
   afterEach(function(done) {
-    if(fileExistsSync('screenshots/')){
-      rimraf.sync('screenshots/');
+    if (utils.dirExistsSync(screenshotFilePath)) {
+      rimraf.sync(screenshotFilePath);
     }
     done();
   });
 
   after(function(done) {
     CommandGlobals.afterEach.call(this, function() {
-      if(fileExistsSync('screenshots/')){
-        rimraf.sync('screenshots/');
+      if (utils.dirExistsSync(screenshotFilePath)) {
+        rimraf.sync(screenshotFilePath);
       }
       done();
     });
@@ -69,7 +72,7 @@ describe('testRunnerScreenshotsOutput', function() {
         enabled: true,
         on_failure: true,
         on_error: true,
-        path: 'screenshots'
+        path: screenshotFilePath
       }
     };
 
@@ -85,12 +88,12 @@ describe('testRunnerScreenshotsOutput', function() {
 
     return NightwatchClient.runTests(testsPath, settings)
       .then(_ => {
-        return readDirPromise('screenshots/sample');
+        return readDirPromise(`${screenshotFilePath}/${moduleName}`);
       })
       .then(files => {
-        assert.ok(files);
-        assert.deepStrictEqual(files.length, 1);
-        files.forEach((val) => assert(val.includes('.png')));
+        assert.ok(Array.isArray(files));
+        assert.strictEqual(files.length, 1);
+        files.forEach((val) => assert.ok(val.endsWith('.png')));
       });
   });
 
@@ -121,7 +124,7 @@ describe('testRunnerScreenshotsOutput', function() {
         enabled: true,
         on_failure: true,
         on_error: true,
-        path: 'screenshots'
+        path: screenshotFilePath
       }
     };
 
@@ -138,11 +141,11 @@ describe('testRunnerScreenshotsOutput', function() {
 
     return NightwatchClient.runTests(testsPath, settings)
       .then(_ => {
-        return readDirPromise('screenshots/sample')
+        return readDirPromise(`${screenshotFilePath}/${moduleName}`)
           .then(files => {
-            assert.ok(files);
+            assert.ok(Array.isArray(files));
             assert.strictEqual(files.length, 1);
-            files.forEach((val) => assert(val.includes('.png')));
+            files.forEach((val) => assert.ok(val.endsWith('.png')));
           });
       });
   });
@@ -174,7 +177,7 @@ describe('testRunnerScreenshotsOutput', function() {
         enabled: true,
         on_failure: true,
         on_error: true,
-        path: 'screenshots'
+        path: screenshotFilePath
       }
     };
 
@@ -195,9 +198,9 @@ describe('testRunnerScreenshotsOutput', function() {
 
     return NightwatchClient.runTests(testsPath, settings)
       .then(_ => {
-        return readDirPromise('screenshots/sample')
+        return readDirPromise(`${screenshotFilePath}/${moduleName}`)
           .catch(err => {
-            assert.ok(err);
+            assert.ok(err instanceof Error);
             assert.ok(err.message.includes('no such file or directory'));
           });
       });
@@ -230,16 +233,16 @@ describe('testRunnerScreenshotsOutput', function() {
         enabled: false,
         on_failure: true,
         on_error: true,
-        path: 'screenshots'
+        path: screenshotFilePath
       }
     };
 
 
     return NightwatchClient.runTests(testsPath, settings)
       .then(_ => {
-        return readDirPromise('screenshots/sample')
+        return readDirPromise(`${screenshotFilePath}/${moduleName}`)
           .catch((err) => {
-            assert.ok(err);
+            assert.ok(err instanceof Error);
             assert.ok(err.message.includes('no such file or directory'));
           });
       });
@@ -272,16 +275,16 @@ describe('testRunnerScreenshotsOutput', function() {
         enabled: true,
         on_failure: false,
         on_error: true,
-        path: 'screenshots'
+        path: screenshotFilePath
       }
     };
 
 
     return NightwatchClient.runTests(testsPath, settings)
       .then(_ => {
-        return readDirPromise('screenshots/sample')
+        return readDirPromise(`${screenshotFilePath}/${moduleName}`)
           .catch((err) => {
-            assert.ok(err);
+            assert.ok(err instanceof Error);
             assert.ok(err.message.includes('no such file or directory'));
           });
       });
@@ -302,12 +305,4 @@ function readDirPromise(dirName) {
   });
 }
 
-// util to replace deprecated fs.existsSync
-function fileExistsSync(path) {
-  try {
-    fs.statSync(path);
-    return true;
-  } catch (e) {
-    return false;
-  }
-}
+

--- a/test/src/runner/testRunnerScreenshotsOutput.js
+++ b/test/src/runner/testRunnerScreenshotsOutput.js
@@ -5,7 +5,6 @@ const common = require('../../common.js');
 const MockServer = require('../../lib/mockserver.js');
 const CommandGlobals = require('../../lib/globals/commands.js');
 const NightwatchClient = common.require('index.js');
-const utils = common.require('utils/index.js');
 const rimraf = require('rimraf');
 
 describe('testRunnerScreenshotsOutput', function() {
@@ -15,26 +14,17 @@ describe('testRunnerScreenshotsOutput', function() {
   before(function(done) {
     this.server = MockServer.init();
     this.server.on('listening', () => {
-      if (utils.dirExistsSync(screenshotFilePath)) {
-        rimraf.sync(screenshotFilePath);
-      }
-      done();
+        rimraf(screenshotFilePath, (error) => {done()});
     });
   });
 
   afterEach(function(done) {
-    if (utils.dirExistsSync(screenshotFilePath)) {
-      rimraf.sync(screenshotFilePath);
-    }
-    done();
+    rimraf(screenshotFilePath, (error) => {done()});
   });
 
   after(function(done) {
     CommandGlobals.afterEach.call(this, function() {
-      if (utils.dirExistsSync(screenshotFilePath)) {
-        rimraf.sync(screenshotFilePath);
-      }
-      done();
+        rimraf(screenshotFilePath, (error) => {done()});
     });
   });
 

